spec/shrinkwrap-and-lock.md

```diff
@@ -0,0 +1,273 @@
+
+Given:
+```
+A (Opt) → B → C
+D → C
+```
+
+Then A and B would be flagged as optional, but C would not be.
+
+Given:
+```
+A (Opt) → B → C
+D → A
+```
+
+Then none would be flagged as optional.
+
+## Installing
+
+The `package-lock.json` or `npm-shrinkwrap.json` describes the exact tree
+that `npm` should create.  Any deviation between the `package.json` and the
+shrinkwrap/lock should result in a warning be issued.  This includes:
+
+* Modules in `package.json` but missing from `npm-shrinkwrap.json` or `package-lock.json`
+* Modules in the `npm-shrinkwrap.json` or `package-lock.json` but missing from the `package.json`.
+* Modules in `package.json` whose specifiers don't match the version in `npm-shrinkwrap.json` or `package-lock.json`.
+
+Warn if the `shrinkwrap-version` in the `npm-shrinkwrap.json` is for a different
+major version than we implement.
+
+Module resolution from shrinkwrap data works as such:
+
+* If install was run with `--resolve-links` and a dependency has a `link`
+  property then a symlink is made using that.  If the version of the
+  destination can not be matched to the shrinkwrap and/or the package.json
+  then a warning will be issued.
+
+* Otherwise, if a `integrity` is available then we try to install it from the cache using it.
+
+If `integrity` is unavailable or we are unable to locate a module from the `integrity` then:
+
+* If `shrinkwrap-version` is set:
+  * Install using the value of `version` and validate the result against the
+    `integrity`.
+* Otherwise, try these in turn and validate the result against the `integrity`:
+  * `resolved`, then `from`, then `version.
+  * `from` can be either `package@specifier` or just `specifier`.
+
+Regardless of how the module is installed the metadata in the installed
+module should be identical to what it would have been if the module were
+installed w/o a shrinkwrap.
+
+## Implied Changes To Other Commands
+
+### `npm rm --save`
+
+Currently if you ask to remove a package that's both a direct and a
+transitive dependency, we'll remove the package from `node_modules` even if
+this results in a broken tree.  This was chosen at the time because we felt
+that users would expect `npm rm pkgname` to be equivalent of
+`rm -rf node_modules/pkgname`.
+
+As you are no longer going to be allowed to put your `node_modules` in a
+state that's not a valid shrinkwrap, this means this behavior is no longer
+valid.  Instead we should follow normal rules, removing it from the
+dependencies for the top level but only removing the module on disk if
+nothing requires it any more.
+
+## Additional fields / Adding new fields
+
+Installers should ignore any field they aren't aware of.  It's not an error
+to have additional properities in the shrinkwrap or lock file.
+
+Installers that want to add new fields should either have one added via RFC
+in the npm issue tracker and an accompanying documentation PR, or should prefix
+it with the name of their project.
```

Comment:
This should specify the format of the prefix - For example, it could be underscore separated (`{project}_{fieldName}`) or colon separated (`{project}:{fieldName}`) or something else altogether.